File: src/main/java/org/taktik/icure/services/external/http/AbstractHttpServlet.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.http;

import java.io.IOException;
import javax.servlet.Servlet;
import javax.servlet.ServletConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class AbstractHttpServlet implements Servlet {
	protected final Logger log = LoggerFactory.getLogger(getClass());

	protected ServletConfig servletConfig;

	@Override
	public void init(ServletConfig servletConfig) throws ServletException {
		this.servletConfig = servletConfig;
	}

	@Override
	public ServletConfig getServletConfig() {
		return servletConfig;
	}

	@Override
	public String getServletInfo() {
		return null;
	}

	@Override
	public void destroy() {
		this.servletConfig = null;
	}

	@Override
	public void service(ServletRequest servletRequest, ServletResponse servletResponse) throws ServletException, IOException {
		if (servletRequest instanceof HttpServletRequest && servletResponse instanceof HttpServletResponse) {
			HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
			HttpServletResponse httpResponse = (HttpServletResponse) servletResponse;
			try {
				handleRequest(httpRequest, httpResponse);
			} catch (ServletException | IOException se) {
				throw se;
			} catch (Exception e) {
				throw new ServletException(e);
			}
		}
	}

	protected abstract void handleRequest(HttpServletRequest httpRequest, HttpServletResponse httpResponse) throws Exception;
}

File: src/main/kotlin/org/taktik/icure/services/external/http/AbstractHttpServlet.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.http

import org.slf4j.LoggerFactory
import java.io.IOException
import javax.servlet.Servlet
import javax.servlet.ServletConfig
import javax.servlet.ServletException
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

abstract class AbstractHttpServlet : Servlet {
    protected val log = LoggerFactory.getLogger(javaClass)
    protected var theServletConfig: ServletConfig? = null
    @Throws(ServletException::class)
    override fun init(servletConfig: ServletConfig) {
        this.theServletConfig = servletConfig
    }

    override fun getServletConfig(): ServletConfig {
        return theServletConfig!!
    }

    override fun getServletInfo(): String? {
        return null
    }

    override fun destroy() {
        theServletConfig = null
    }

    @Throws(ServletException::class, IOException::class)
    override fun service(servletRequest: ServletRequest, servletResponse: ServletResponse) {
        if (servletRequest is HttpServletRequest && servletResponse is HttpServletResponse) {
            try {
                handleRequest(servletRequest, servletResponse)
            } catch (se: ServletException) {
                throw se
            } catch (se: IOException) {
                throw se
            } catch (e: Exception) {
                throw ServletException(e)
            }
        }
    }

    @Throws(Exception::class)
    protected abstract fun handleRequest(httpRequest: HttpServletRequest?, httpResponse: HttpServletResponse?)
}
